### ISSUE TITLE
Problems with Stopwatch resolution on macOS

### ISSUE THREAD

**Issue by AndreyAkinshin** [dotnet/corefx#30391]
Consider the following program:
```xml
<Project Sdk="Microsoft.NET.Sdk">
    <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFrameworks>netcoreapp1.1;netcoreapp2.1</TargetFrameworks>
    </PropertyGroup>
</Project>
```
```cs
class Program
{
    static void Main()
    {
        Console.WriteLine(Stopwatch.GetTimestamp());
        Console.WriteLine(Stopwatch.GetTimestamp());
        Console.WriteLine(Stopwatch.GetTimestamp());
        Console.WriteLine(Stopwatch.GetTimestamp());
        Console.WriteLine(Stopwatch.GetTimestamp());
    }
}
```
Now let's run it on macOS 10.13.4 for both target frameworks. I tried 3 different machines and my results always look like this:
```
$ dotnet run -f netcoreapp1.1
2385586752983342
2385586768343646
2385586768360549
2385586768364474
2385586768368137
$ dotnet run -f netcoreapp2.1
3958972378494000
3958972381430000
3958972381447000
3958972381451000
3958972381455000
```
As you can see, the stopwatch resolution on .NET Core 2.1 is 1000ns. I have many different benchmarks which confirm this situation. Here is one of them:

```cs
var timestamps = new long[10000];
for (int i = 0; i < timestamps.Length; i++)
    timestamps[i] = Stopwatch.GetTimestamp();
var minDelta = long.MaxValue;
for (int i = 0; i < timestamps.Length - 1; i++)
    if (timestamps[i] != timestamps[i + 1])
        minDelta = Math.Min(minDelta, timestamps[i + 1] - timestamps[i]);
Console.WriteLine(minDelta);
```

Results:

```
$ dotnet run -f netcoreapp1.1
31
$ dotnet run -f netcoreapp2.1
1000
```

Here is the relevant code (https://github.com/dotnet/corefx/blob/v2.1.0/src/Native/Unix/System.Native/pal_time.cpp#L81):
```cpp
extern "C" int32_t SystemNative_GetTimestamp(uint64_t* timestamp)
{
    assert(timestamp);

#if HAVE_CLOCK_MONOTONIC
    struct timespec ts;
    int result = clock_gettime(CLOCK_MONOTONIC, &ts);
    assert(result == 0); // only possible errors are if MONOTONIC isn't supported or &ts is an invalid address
    (void)result; // suppress unused parameter warning in release builds
    *timestamp = (static_cast<uint64_t>(ts.tv_sec) * SecondsToNanoSeconds) + static_cast<uint64_t>(ts.tv_nsec);
    return 1;

#elif HAVE_MACH_ABSOLUTE_TIME
    *timestamp = mach_absolute_time();
    return 1;

#else
    struct timeval tv;
    if (gettimeofday(&tv, NULL) == 0)
    {
        *timestamp = (static_cast<uint64_t>(tv.tv_sec) * SecondsToMicroSeconds) + static_cast<uint64_t>(tv.tv_usec);
        return 1;
    }
    else
    {
        *timestamp = 0;
        return 0;
    }

#endif
}
```
I didn't find any relevant changes in the source code between .NET Core 1.x and .NET Core 2.x.
Next, I tried a pure C program:
```c
#include <stdio.h>
#include <time.h>

int main() {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    printf("ts.tv_sec  = %ld\n", ts.tv_sec);
    printf("ts.tv_nsec = %ld\n", ts.tv_nsec);
    return 0;
}
```
The results:
```
ts.tv_sec  = 3855196
ts.tv_nsec = 495855000

ts.tv_sec  = 3855196
ts.tv_nsec = 976426000

ts.tv_sec  = 3855197
ts.tv_nsec = 437808000

ts.tv_sec  = 3855197
ts.tv_nsec = 911397000

ts.tv_sec  = 3855199
ts.tv_nsec = 293822000
```
Thus, native `CLOCK_MONOTONIC` also has the 1000ns resolution.
Also I found a similar golang issue: https://github.com/golang/go/issues/22037 (probably the problem is related to macOS version).

Open questions:
* What the difference between .NET Core 1.x and .NET Core 2.x? Why `Stopwatch.GetTimestamp()` is affected?
* How can we improve the Stopwatch resolution on macOS?

/cc @stephentoub

**Comment by jkotas** [comment]
macOS should take the `HAVE_MACH_ABSOLUTE_TIME` path. It is the native macOS API that will be fastest and most accurate.

What likely happened:
- .NET Core 1.x was compiled against old macOS SDK and we took `mach_absolute_time` path as expected
- Newer macOS SDK added of `CLOCK_MONOTONIC` for better compatibility with other Unix system
- .NET Core 2.x is compiled against newer macOS SDK and picking up `CLOCK_MONOTONIC` by accident

**Comment by AndreyAkinshin** [comment]
@jkotas,

> .NET Core 1.x was compiled against old macOS SDK 
> .NET Core 2.x is compiled against newer macOS SDK

It sounds believable.

By the way, the same code in Mono (see https://github.com/mono/mono/blob/mono-5.12.0.226/mono/utils/mono-time.c#L208) try to use `mach_absolute_time` first.
The order of `mach_absolute_time` and `clock_gettime` was changed in https://github.com/mono/mono/pull/3165 (because of the macOS 10.12 SDK; `CLOCK_MONOTONIC` was introduced in this version).

I'm ready to send a pull request and make the same change for .NET Core (if it sounds OK for you).

**Comment by jkotas** [comment]
PRs to fix this are welcomed!

**Comment by stephentoub** [comment]
Thanks, @AndreyAkinshin.  This affects more than just Stopwatch, and should be addressed in both coreclr:

https://github.com/dotnet/coreclr/blob/549c9960a8edcbe3930639e316616d35b22bca25/src/pal/src/misc/time.cpp#L205-L220

https://github.com/dotnet/coreclr/blob/549c9960a8edcbe3930639e316616d35b22bca25/src/pal/src/misc/time.cpp#L267-L279

https://github.com/dotnet/coreclr/blob/549c9960a8edcbe3930639e316616d35b22bca25/src/pal/src/misc/time.cpp#L341-L367

and corefx:

https://github.com/dotnet/corefx/blob/b3add58d237c1089e1a0bfd106ff510e8c23242f/src/Native/Unix/System.Native/pal_time.c#L45-L73

https://github.com/dotnet/corefx/blob/b3add58d237c1089e1a0bfd106ff510e8c23242f/src/Native/Unix/System.Native/pal_time.c#L85-L96

**Comment by AndreyAkinshin** [comment]
@stephentoub thanks for the links! I will update all usages of `CLOCK_MONOTONIC` in both repositories.

**Comment by karelz** [comment]
Fixed in 2.1 servicing by PR #30459. Will be part of 2.1.3 release.
